File: popularity-sort-uses-stale-total-sales/python/recount_total_sales.py
"""Recount WooCommerce's total_sales product meta from real, paid orders.

The "Popularity" catalog sort orders products by the total_sales number stored
on each product. WooCommerce core only bumps that number through its own order
status hooks, so it drifts from reality whenever orders are imported straight
into the database, a status is changed by another plugin or a direct SQL
update, or a refund and cancellation never decrements it back down. This walks
paid orders in a lookback window, sums real quantities per product (minus
refunded quantities), compares that to the stored total_sales, and corrects
any product whose number is wrong. Read only by default. Run on a schedule.
"""
import os
import logging
import requests
from collections import defaultdict
from requests.auth import HTTPBasicAuth

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger("recount_total_sales")

WOO_URL = os.environ["WOO_STORE_URL"].rstrip("/")
AUTH = HTTPBasicAuth(os.environ["WOO_CONSUMER_KEY"], os.environ["WOO_CONSUMER_SECRET"])
LOOKBACK_DAYS = int(os.environ.get("LOOKBACK_DAYS", "365"))
DRY_RUN = os.environ.get("DRY_RUN", "true").lower() == "true"

# WooCommerce counts a sale once the order reaches one of these statuses.
SALE_COUNTED_STATUSES = {"processing", "completed"}


def decide(stored_total_sales, real_total_sales):
    """Pure decision: should we rewrite a product's total_sales meta?

    stored_total_sales: the int currently saved on the product (what Popularity sorts by).
    real_total_sales: the int we computed from real order line items and refunds.
    Returns a tuple of (action, reason). action is one of "fix", "skip".
    """
    try:
        stored = int(stored_total_sales)
    except (TypeError, ValueError):
        stored = 0
    real = max(0, int(real_total_sales))
    if stored == real:
        return ("skip", "total_sales already correct")
    return ("fix", f"stored {stored}, real {real}")


def net_quantity(line_item):
    """Quantity actually sold for one order line item, in whole units.

    A negative qty on a refund line item cancels out units from the original
    order line item for the same product.
    """
    try:
        return int(line_item.get("quantity") or 0)
    except (TypeError, ValueError):
        return 0


def paid_orders(after_iso):
    page = 1
    while True:
        r = requests.get(
            f"{WOO_URL}/wp-json/wc/v3/orders",
            params={
                "status": "processing,completed",
                "after": after_iso,
                "per_page": 100,
                "page": page,
            },
            auth=AUTH, timeout=30,
        )
        r.raise_for_status()
        batch = r.json()
        if not batch:
            return
        for order in batch:
            yield order
        page += 1


def refund_line_items(order_id):
    r = requests.get(
        f"{WOO_URL}/wp-json/wc/v3/orders/{order_id}/refunds",
        auth=AUTH, timeout=30,
    )
    r.raise_for_status()
    for refund in r.json():
        for item in refund.get("line_items", []):
            yield item


def real_sales_by_product(lookback_days):
    """Sum real, paid quantity per product_id across the lookback window."""
    import datetime

    after = (datetime.date.today() - datetime.timedelta(days=lookback_days)).isoformat() + "T00:00:00"
    totals = defaultdict(int)
    for order in paid_orders(after):
        for item in order.get("line_items", []):
            product_id = item.get("product_id")
            if not product_id:
                continue
            totals[product_id] += net_quantity(item)
        for item in refund_line_items(order["id"]):
            product_id = item.get("product_id")
            if not product_id:
                continue
            # Refund line items carry a negative quantity already.
            totals[product_id] += net_quantity(item)
    return totals


def get_product(product_id):
    r = requests.get(f"{WOO_URL}/wp-json/wc/v3/products/{product_id}", auth=AUTH, timeout=30)
    if r.status_code == 404:
        return None
    r.raise_for_status()
    return r.json()


def write_total_sales(product_id, real_total_sales):
    requests.put(
        f"{WOO_URL}/wp-json/wc/v3/products/{product_id}",
        json={"total_sales": str(real_total_sales)},
        auth=AUTH, timeout=30,
    ).raise_for_status()


def run():
    fixed = 0
    totals = real_sales_by_product(LOOKBACK_DAYS)
    for product_id, real_total_sales in totals.items():
        product = get_product(product_id)
        if product is None:
            log.warning("Product %s has sales but no longer exists, skipping", product_id)
            continue
        action, reason = decide(product.get("total_sales"), real_total_sales)
        if action == "skip":
            continue
        log.info("Product %s: %s. %s", product_id, reason, "would fix" if DRY_RUN else "fixing")
        if not DRY_RUN:
            write_total_sales(product_id, real_total_sales)
        fixed += 1
    log.info("Done. %d product(s) %s.", fixed, "to fix" if DRY_RUN else "fixed")


if __name__ == "__main__":
    run()

File: popularity-sort-uses-stale-total-sales/node/recount-total-sales.js
/**
 * Recount WooCommerce's total_sales product meta from real, paid orders.
 *
 * The "Popularity" catalog sort orders products by the total_sales number
 * stored on each product. WooCommerce core only bumps that number through
 * its own order status hooks, so it drifts from reality whenever orders are
 * imported straight into the database, a status is changed by another
 * plugin or a direct SQL update, or a refund and cancellation never
 * decrements it back down. This walks paid orders in a lookback window,
 * sums real quantities per product (minus refunded quantities), compares
 * that to the stored total_sales, and corrects any product whose number is
 * wrong. Read only by default. Run on a schedule.
 *
 * Guide: https://www.allanninal.dev/woocommerce/popularity-sort-uses-stale-total-sales/
 */
import { pathToFileURL } from "node:url";

const WOO_URL = (process.env.WOO_STORE_URL || "https://example.com").replace(/\/$/, "");
const AUTH = "Basic " + Buffer.from(
  `${process.env.WOO_CONSUMER_KEY || "ck_dummy"}:${process.env.WOO_CONSUMER_SECRET || "cs_dummy"}`
).toString("base64");
const LOOKBACK_DAYS = Number(process.env.LOOKBACK_DAYS || 365);
const DRY_RUN = (process.env.DRY_RUN || "true").toLowerCase() === "true";

/**
 * Pure decision: should we rewrite a product's total_sales meta?
 * stored: the value currently saved on the product (what Popularity sorts by).
 * real: the number we computed from real order line items and refunds.
 * Returns [action, reason] where action is "fix" or "skip".
 */
export function decide(storedTotalSales, realTotalSales) {
  const stored = Number.isFinite(Number(storedTotalSales)) ? Math.trunc(Number(storedTotalSales)) : 0;
  const real = Math.max(0, Math.trunc(realTotalSales || 0));
  if (stored === real) return ["skip", "total_sales already correct"];
  return ["fix", `stored ${stored}, real ${real}`];
}

/**
 * Quantity actually sold for one order line item, in whole units.
 * A negative qty on a refund line item cancels out units from the
 * original order line item for the same product.
 */
export function netQuantity(lineItem) {
  const qty = Number(lineItem && lineItem.quantity);
  return Number.isFinite(qty) ? Math.trunc(qty) : 0;
}

async function woo(path, options = {}) {
  const res = await fetch(`${WOO_URL}/wp-json/wc/v3${path}`, {
    ...options,
    headers: { "Content-Type": "application/json", Authorization: AUTH, ...(options.headers || {}) },
  });
  if (res.status === 404) return null;
  if (!res.ok) throw new Error(`Woo ${path} returned ${res.status}`);
  return res.json();
}

async function* paidOrders(afterIso) {
  let page = 1;
  while (true) {
    const batch = await woo(
      `/orders?status=processing,completed&after=${afterIso}&per_page=100&page=${page}`
    );
    if (!batch || !batch.length) return;
    for (const order of batch) yield order;
    page++;
  }
}

async function refundLineItems(orderId) {
  const refunds = (await woo(`/orders/${orderId}/refunds`)) || [];
  const items = [];
  for (const refund of refunds) {
    for (const item of refund.line_items || []) items.push(item);
  }
  return items;
}

async function realSalesByProduct(lookbackDays) {
  const after = new Date(Date.now() - lookbackDays * 86400000).toISOString();
  const totals = new Map();
  for await (const order of paidOrders(after)) {
    for (const item of order.line_items || []) {
      if (!item.product_id) continue;
      totals.set(item.product_id, (totals.get(item.product_id) || 0) + netQuantity(item));
    }
    for (const item of await refundLineItems(order.id)) {
      if (!item.product_id) continue;
      // Refund line items carry a negative quantity already.
      totals.set(item.product_id, (totals.get(item.product_id) || 0) + netQuantity(item));
    }
  }
  return totals;
}

async function writeTotalSales(productId, realTotalSales) {
  await woo(`/products/${productId}`, {
    method: "PUT",
    body: JSON.stringify({ total_sales: String(realTotalSales) }),
  });
}

export async function run() {
  let fixed = 0;
  const totals = await realSalesByProduct(LOOKBACK_DAYS);
  for (const [productId, realTotalSales] of totals) {
    const product = await woo(`/products/${productId}`);
    if (!product) {
      console.warn(`Product ${productId} has sales but no longer exists, skipping`);
      continue;
    }
    const [action, reason] = decide(product.total_sales, realTotalSales);
    if (action === "skip") continue;
    console.log(`Product ${productId}: ${reason}. ${DRY_RUN ? "would fix" : "fixing"}`);
    if (!DRY_RUN) await writeTotalSales(productId, realTotalSales);
    fixed++;
  }
  console.log(`Done. ${fixed} product(s) ${DRY_RUN ? "to fix" : "fixed"}.`);
}

if (import.meta.url === pathToFileURL(process.argv[1]).href) {
  run().catch((err) => { console.error(err); process.exit(1); });
}
